### DISCUSSION TITLE
Potential bug in automerging on GitHub

### DISCUSSION THREAD

**Discussion by nwalters512** [renovatebot/renovate#18593] [Request Help] [answered]
I've tried to enable automerging in one of my repos (I'm self-hosting Renovate). After a fair amount of debugging, I think I've discovered an issue with the way that auto-merging occurs for existing PRs, but I'd like to confirm it first here before I open an issue. I don't have a great understanding of Renovate's internals, so it's possible this is actually working as intended.

It looks like Renovate will always try to recreate a PR, even if one already exists. On GitHub, that results in an error response, which is handled here:

https://github.com/renovatebot/renovate/blob/0e193aded53b409d84c3403400fa5fce8d314200/lib/workers/repository/update/pr/index.ts#L374-L383

That means that in `processBranch`, we hit this case:

https://github.com/renovatebot/renovate/blob/0e193aded53b409d84c3403400fa5fce8d314200/lib/workers/repository/update/branch/index.ts#L737-L744

That early return means we never actually attempt to auto-merge later on:

https://github.com/renovatebot/renovate/blob/0e193aded53b409d84c3403400fa5fce8d314200/lib/workers/repository/update/branch/index.ts#L803

---

As an asside, I think the following check is broken:

https://github.com/renovatebot/renovate/blob/0e193aded53b409d84c3403400fa5fce8d314200/lib/workers/repository/update/pr/index.ts#L375

When I enable debug logs, I see that the response from GitHub looks like the following:

```json
"body": {
  "message": "Validation Failed",
  "errors": [
    {
      "resource": "PullRequest",
      "code": "custom",
      "message": "A pull request already exists for XXX:YYY"
    }
  ],
  "documentation_url": "https://docs.github.com/rest/reference/pulls#create-a-pull-request"
},
```

Notice the difference in capitalization: `Validation Failed` in the actual error message, but `Validation failed` in Renovate's check.

That said, I don't think that currently manifests in a difference in behavior, as the same message is returned regardless.

**Comment by rarkins** [comment, marked as answer]
> It looks like Renovate will always try to recreate a PR, even if one already exists.

This is not the case. Renovate will first try to find an existing PR for the branch. If one is not found, then it tries to recreate.

The most common reason why you might be seeing this is because you have the wrong `username` configured for the bot. How exactly are you running it, e.g. via GitHub Action, as an App, etc?

**Reply by nwalters512** [reply]
Ah! That makes sense. I saw this behavior when I was testing the CLI locally via a GitHub PAT. Normally Renovate runs on a schedule as a machine user, so the existing PRs were associated with that other user. It seems like Renovate looks up existing PRs by the username that Renovate is executing as?

**Reply by viceice** [reply]
yes
